### PR TITLE
feat: (Platform) introduce new Table component 

### PR DESCRIPTION
**Pull request by olshansky**
#### Please provide a link to the associated issue.
https://github.com/SAP/fundamental-ngx/issues/3252
#### Please provide a brief summary of this pull request.
The Grid Table is a table component for displaying large data sets.

- Basic table toolbar
- Basic column definitions
- Custom column definitions
- Single item/row selection
- Multiple item/row selection
#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] Documentation Examples

### CODE REVIEW

**Review by Betrozov — APPROVED** [review]

**Review by KevinOkamoto — APPROVED** [review]
Looks good!

**Comment by mikerodonnell89** [comment]
The "single row selection" example has an extra empty column

**Comment by olshansky** [comment]
> The "single row selection" example needs the checkboxes

@mikerodonnell89 According to [the design guidelines](https://sapui5.hana.ondemand.com/#/entity/sap.ui.table.Table/sample/sap.ui.table.sample.Selection) we do not need checkboxes
image

**Comment by dimamarksman** [comment]
Looks like we don't support Up/Down arrow navigation, but according to the spec we should.

image

**Comment by olshansky** [comment]
> Looks like we don't support Up/Down arrow navigation, but according to the spec we should.

@dimamarksman On one of the latest specs demo related to the table _(do not really remember styles or ngx spec)_ was shown "focusable" feature (when the cells could be focusable and navigatable elements). The same I've seen [here](https://sapui5.hana.ondemand.com/#/entity/sap.ui.table.Table/sample/sap.ui.table.sample.Selection).
That's why I postponed it and left a [comment](https://github.com/SAP/fundamental-ngx/pull/3597/files#diff-40c500ce441f2822f12fe43cf11ead5b2160aee2c7c0a53921b3dce6feca4816R252) for the feature. 
I hope for now it enough to have an ability to navigate the checkboxes by `Tab/Shift+Tab` 🙃

**Comment by dimamarksman** [comment]
> > Looks like we don't support Up/Down arrow navigation, but according to the spec we should.
> 
> @dimamarksman On one of the latest specs demo related to the table _(do not really remember styles or ngx spec)_ was shown "focusable" feature (when the cells could be focusable and navigatable elements). The same I've seen [here](https://sapui5.hana.ondemand.com/#/entity/sap.ui.table.Table/sample/sap.ui.table.sample.Selection).
> That's why I postponed it and left a [comment](https://github.com/SAP/fundamental-ngx/pull/3597/files#diff-40c500ce441f2822f12fe43cf11ead5b2160aee2c7c0a53921b3dce6feca4816R252) for the feature.
> I hope for now it enough to have an ability to navigate the checkboxes by `Tab/Shift+Tab`

So we are going to add this ability later. OK

File: libs/platform/src/lib/components/table/table.component.ts [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+    rows: SelectableRow[] = [];
+
+    /** @hidden */
+    checkedAll = false;
+
+    /** @hidden */
+    checked = [];
+
+    /** @hidden */
+    unchecked = [];
```
Comment (dimamarksman) [inline libs/platform/src/lib/components/table/table.component.ts:137]: Do we need to keep it public?

File: libs/platform/src/lib/components/table/table.component.ts [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+
+    /** @hidden */
+    contentDensityOptions = ContentDensity;
+
+    /** @hidden */
+    selectionModeOptions = SelectionMode;
```
Comment (dimamarksman) [inline libs/platform/src/lib/components/table/table.component.ts:143]: Sunce it doesn't belong to public API can we add a prefix `_`?

File: libs/platform/src/lib/domain/data-source.ts [hunk 3]
```diff
@@ -230,3 +230,30 @@ export class ListDataSource<T> extends ComboBoxDataSource<T> {
     }
 }
 
+export class TableDataSource<T> extends ListDataSource<T> implements DataSource<T> {
+    readonly MaxLimit = 10;
```
Comment (dimamarksman) [inline libs/platform/src/lib/domain/data-source.ts:234]: if intention is to highlight that it is a constant then it should be `MAX_LIMIT`.
`MaxLimit` looks like a reference to a `class` or `type`

**Review by dimamarksman — APPROVED** [review]
Looks good

File: apps/docs/src/app/platform/api-files.ts [hunk 4]
```diff
@@ -20,6 +20,12 @@ export const API_FILES = {
     searchField: ['SearchFieldComponent'],
     select: ['SelectPlatformComponent'],
     splitMenuButton: ['SplitMenuButtonComponent'],
+    table: [
+        'TableComponent',
+        'TableToolbarComponent',
+        'TableToolbarActionsComponent',
+        'TableColumnComponent'
+    ],
```
Comment (dimamarksman) [inline apps/docs/src/app/platform/api-files.ts:28]: We need to add these directives as well:
``` 
  FdpTableCell,
  FdpCellDef,
  FdpTableHeader,
  FdpHeaderCellDef
```

Comment (olshansky) [reply apps/docs/src/app/platform/api-files.ts:28]: Thanks, you absolutely right

File: libs/platform/src/lib/components/table/table.component.ts [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+} from '@angular/core';
+
+import { isObservable, Subject, Subscription } from 'rxjs';
+import { takeUntil } from 'rxjs/operators';
+
+import { applyCssClass, CssClassBuilder, KeyUtil } from '@fundamental-ngx/core';
+import { ContentDensity, SelectionMode } from './enums';
+import { TableColumnComponent } from './table-column/table-column.component';
+import { TableToolbarComponent } from './table-toolbar/table-toolbar.component';
+import { ArrayTableDataSource, isDataSource, ObservableTableDataSource, TableDataSource } from '../../../';
```
Comment (JKMarkowski) [inline libs/platform/src/lib/components/table/table.component.ts:28]: Can you add direct path there?

**Review by JKMarkowski — APPROVED** [review]
Hi @olshansky , check my one comment I added, rest LGTM.
Also I'm wondering how about a rest of table API things like `interractive`, `no borders`, or 'pop-in'?

**Comment by olshansky** [comment]
> Also I'm wondering how about a rest of table API things like `interractive`, `no borders`, or 'pop-in'?

Hi @JKMarkowski. Thanks for the review.
Nothing from these wasn't mentioned in [Platform: Table - Phase 1 spec](https://app.gitbook.com/@fundamental-ngx/s/docs/platform-table-phase-1), so I hope it would be included in the next phases specs and implementation

**Pull request merged**